src/session/download.c

```diff
@@ -999,31 +1001,17 @@ create_download_file(struct terminal *term, char *fi,
 static char *
 get_temp_name(struct uri *uri)
 {
-	struct string name;
 	char *extension;
-	/* FIXME
-	 * We use tempnam() here, which is unsafe (race condition), for now.
-	 * This should be changed at some time, but it needs an in-depth work
-	 * of whole download code. --Zas */
-	char *nm = tempnam(NULL, ELINKS_TEMPNAME_PREFIX);
-
-	if (!nm) return NULL;
-
-	if (!init_string(&name)) {
-		free(nm);
-		return NULL;
-	}
-
-	add_to_string(&name, nm);
-	free(nm);
+	char *nm;
 
 	extension = get_extension_from_uri(uri);
-	if (extension) {
-		add_shell_safe_to_string(&name, extension, strlen(extension));
-		mem_free(extension);
-	}
+	if (!extension)
+		extension = strdup("");
```

Comment:
mem_alloc, mem_calloc, mem_realloc, stracpy, memacpy -> mem_free. If you allocate memory with one of these function use mem_free to release memory.
malloc, calloc, realloc, strdup -> free . For these function use free.
Do not mix them. With --enable-fastmem both groups are the same, but for other options not.
Especially --enable-debug.

In above code stracpy must be used instead of strdup.
Earlier in tempname, strdup was used, because tempnam also allocated memory with malloc.